sql: TestStatementTimeoutForSchemaChangeCommit failed

sql.TestStatementTimeoutForSchemaChangeCommit [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17444037?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17444037?buildTab=artifacts#/) on release-24.3 @ [8921bdb44f0148eb7dbca26ed658b7d83251038d](https://github.com/cockroachdb/cockroach/commits/8921bdb44f0148eb7dbca26ed658b7d83251038d):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestStatementTimeoutForSchemaChangeCommit (13m11s)
	TestStatementTimeoutForSchemaChangeCommit/implicitTxn=true (13m11s)

```
Stack: 

```
goroutine 158294 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* runtime/pprof.Do({0x82d6ae8?, 0xc8f80c0?}, {{0x40017d44a0?, 0x4009122f18?, 0x4009122f68?}}, 0x4009122f98)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x6a02405?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:397 +0x5c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 156164
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:389 +0xe4
* 
* goroutine 156164 [semacquire, 2 minutes]:
* sync.runtime_Semacquire(0x4004e25880?)
* 	GOROOT/src/runtime/sema.go:62 +0x2c
* sync.(*WaitGroup).Wait(0x4001cf84b0)
* 	GOROOT/src/sync/waitgroup.go:116 +0x74
* github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).stopServers(0x400742aa08, {0x82d9d00, 0xc8f80c0})
* 	github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:155 +0x234
* github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start.func3()
* 	github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:448 +0x30
* github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0x401103db90?)
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:103 +0x24
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0x401103db90, {0x82d6ae8, 0xc8f80c0})
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:563 +0x108
* runtime.Goexit()
* 	GOROOT/src/runtime/panic.go:626 +0x60
* testing.(*common).FailNow(0x4007737040)
* 	GOROOT/src/testing/testing.go:1005 +0x48
* github.com/stretchr/testify/require.ErrorContains({0x8296398, 0x4007737040}, {0x827abc0, 0x400fa645a0}, {0x4004c100c0, 0x31}, {0x0, 0x0, 0x0})
* 	github.com/stretchr/testify/require/external/com_github_stretchr_testify/require/require.go:332 +0xdc
* github.com/cockroachdb/cockroach/pkg/sql_test.TestStatementTimeoutForSchemaChangeCommit.func1(0x4007737040)
* 	github.com/cockroachdb/cockroach/pkg/sql_test/pkg/sql/run_control_test.go:1021 +0x484
* testing.tRunner(0x4007737040, 0x400b562c20)
* 	GOROOT/src/testing/testing.go:1689 +0xec
* created by testing.(*T).Run in goroutine 155711
* 	GOROOT/src/testing/testing.go:1742 +0x318
* 
* goroutine 156670 [select, 2 minutes]:
* github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:501 +0x140
* created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs in goroutine 156164
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:472 +0x5c
* 
* goroutine 156200 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x82d9fa0?, 0x400ab44030?})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:399 +0x80
* runtime/pprof.Do({0x82d6ae8?, 0xc8f80c0?}, {{0x40017d44a0?, 0x4007eae718?, 0x4007eae768?}}, 0x4003183f98)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x6941745?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:397 +0x5c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 156164
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:389 +0xe4
* 
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatementTimeoutForSchemaChangeCommit.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-43582

--- a/pkg/sql/run_control_test.go
+++ b/pkg/sql/run_control_test.go
@@ -1016,6 +1016,10 @@ func TestStatementTimeoutForSchemaChangeCommit(t *testing.T) {
 				require.NoError(t, err)
 				// Test implicit transactions first.
 				blockSchemaChange.Swap(true)
+				defer func() {
+					close(waitForTimeout)
+					blockSchemaChange.Swap(false)
+				}()
 				if implicitTxn {
 					_, err := conn.DB.ExecContext(ctx, "ALTER TABLE t1 ADD COLUMN j INT DEFAULT 32")
 					require.ErrorContains(t, err, sqlerrors.QueryTimeoutError.Error())
@@ -1030,8 +1034,6 @@ func TestStatementTimeoutForSchemaChangeCommit(t *testing.T) {
 					err = txn.Commit()
 					require.NoError(t, err)
 				}
-				close(waitForTimeout)
-				blockSchemaChange.Swap(false)
 			})
 	}
 }

